src/compile.rs

```diff
@@ -40,191 +43,339 @@ impl Compiler {
-            Expr::Repeat { e, r: Repeater::ZeroOrMore, greedy } => {
-                let j1 = self.insts.len();
-                let split = self.empty_split();
-                let j2 = self.insts.len();
-                try!(self.c(*e));
-                let jmp = self.empty_jump();
-                let j3 = self.insts.len();
-
-                self.set_jump(jmp, j1);
-                if greedy {
-                    self.set_split(split, j2, j3);
-                } else {
-                    self.set_split(split, j3, j2);
-                }
+            Repeater::Range { min, max: Some(max) } => {
+                self.c_repeat_range(expr, greedy, min, max)
             }
-            Expr::Repeat { e, r: Repeater::OneOrMore, greedy } => {
-                let j1 = self.insts.len();
-                try!(self.c(*e));
-                let split = self.empty_split();
-                let j2 = self.insts.len();
-
-                if greedy {
-                    self.set_split(split, j1, j2);
-                } else {
-                    self.set_split(split, j2, j1);
-                }
+        }
+    }
+
+    fn c_repeat_zero_or_one(
+        &mut self,
+        expr: &Expr,
+        greedy: bool,
+    ) -> CompileResult {
+        let split = self.push_split_hole();
+        let goto1 = self.insts.len();
+        let hole = try!(self.c(expr));
+        let goto2 = self.insts.len();
+
+        if greedy {
+            self.fill_split(split, Some(goto1), Some(goto2));
+        } else {
+            self.fill_split(split, Some(goto2), Some(goto1));
+        }
+        Ok(hole)
+    }
+
+    fn c_repeat_zero_or_more(
+        &mut self,
+        expr: &Expr,
+        greedy: bool,
+    ) -> CompileResult {
+        let goto_split = self.insts.len();
+        let split = self.push_split_hole();
+        let goto_rep_expr = self.insts.len();
+        let hole_rep_expr = try!(self.c(expr));
+
+        self.fill(hole_rep_expr, goto_split);
+        Ok(if greedy {
+            self.fill_split(split, Some(goto_rep_expr), None)
+        } else {
+            self.fill_split(split, None, Some(goto_rep_expr))
+        })
+    }
+
+    fn c_repeat_one_or_more(
+        &mut self,
+        expr: &Expr,
+        greedy: bool,
+    ) -> CompileResult {
+        let goto_rep_expr = self.insts.len();
+        let hole_rep_expr = try!(self.c(expr));
+        self.fill_to_next(hole_rep_expr);
+        let split = self.push_split_hole();
+
+        Ok(if greedy {
+            self.fill_split(split, Some(goto_rep_expr), None)
+        } else {
+            self.fill_split(split, None, Some(goto_rep_expr))
+        })
+    }
+
+    fn c_repeat_range_min_or_more(
+        &mut self,
+        expr: &Expr,
+        greedy: bool,
+        min: u32,
+    ) -> CompileResult {
+        let min = u32_to_usize(min);
+        let hole = try!(self.c_concat(iter::repeat(expr).take(min)));
+        self.fill_to_next(hole);
+        self.c_repeat_zero_or_more(expr, greedy)
```

Comment:
If `min >= 1`, you would get a smaller machine by repeating `min - 1` times and then doing `repeat_one_or_more`